### PR TITLE
Math expression converter

### LINKED ISSUE #237: [Enhancement] ArithmeticConverter
## Summary

Allow performing arithmetic operations in XAML.

https://github.com/Apress/pro-wpf-4.5-in-csharp/blob/master/Pro%20WPF%204.5%20in%20C%23/Chapter15/Animation/ArithmeticConverter.cs

### PR DESCRIPTION
**Pull request by sattew**
### Description of Change ###

So, I added converter (main and multi version) to calculate an expression in XAML. As @AndreiMisiukevich wrote I realized the RPN algorithm to solve it.

#### Allowed operations:  ####
- \+
- \-
- /
- \*
- %
- ^ or pow(double, double) 
- abs(double)
- acos(double)
- asin(double)
- atan(double)
- atan2(double, double)
- ceiling(double)
- cos(double)
- cosh(double)
- exp(double)
- floor(double)
- ieeeremainder(double)
- log(double, double)
- logn(double)
- max(double)
- min(double)
- round(double, int)
- sign(double)
- sin(double)
- sinh(double)
- sqrt(double)
- tan(double)
- tanh(double)
- truncate
- pi
- e

All operators are similar to the methods of the Math class.

User's variables what be added via converter would be notice like

- **x** or **x0** — first arg
- **x1** — second
- ...
- **xN-1** — other

PS. The negative func (or constants/variables) like "-cos(30)" or "-x", "-pi" doesn't implement, but it will work if replace it to "-1\*cos(30)" or "-1\*x".
And I set all classes (except converters) to internal because i think it's correct ^_^
Also I don't have any Apple product and I haven't tested it. I'm almost sure it will work on these devices.

### Issues ###

- Closed #237

### API Changes ###

Added: 

```csharp
public class MathExpressionConverter : ValueConverterExtension, IValueConverter
```
```csharp
public class MultiMathExpressionConverter : MultiValueConverterExtension, IMultiValueConverter
```
### Behavioral Changes ###

None

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [ ] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard
- [x] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/MathExpressionConverter.shared.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+﻿using System;
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/MathExpressionConverter.shared.cs:1]: Please create a folder **MathExpressionConverter** and put all math files there.
Keep in mind, that the namespace should be **Xamarin.CommunityToolkit.Converters** for all files in that folder.

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathOperatorPrecedence.shared.cs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+﻿namespace Xamarin.CommunityToolkit.Core
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathOperatorPrecedence.shared.cs:1]: please move to **MathExpressionConverter** folder and amend the namespace

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathOperator.shared.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+
+namespace Xamarin.CommunityToolkit.Core
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathOperator.shared.cs:3]: please move to **MathExpressionConverter** folder and amend the namespace

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathExpression.shared.cs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+			};
+
+			if (this.arguments.Count > 0)
+			{
+				operators.Add(new MathOperator("x", 0, MathOperatorPrecedence.Constant, _ => this.arguments[0]));
+			}
+
+			for (var i = 0; i < this.arguments.Count; i++)
+			{
+				var i1 = i;
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathExpression.shared.cs:71]: can you call it "index" ?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathExpression.shared.cs [hunk 5]
```diff
@@ -0,0 +1,209 @@
+﻿#nullable enable
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Xamarin.CommunityToolkit.Core
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathExpression.shared.cs:9]: please move to **MathExpressionConverter** folder and amend the namespace

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathExpression.shared.cs [hunk 6]
```diff
@@ -0,0 +1,209 @@
+﻿#nullable enable
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathExpression.shared.cs:1]: nullable is enabled by default, I believe

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/MultiMathExpressionConverter.shared.cs [hunk 7]
```diff
@@ -0,0 +1,57 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using Xamarin.CommunityToolkit.Core;
+using Xamarin.CommunityToolkit.Extensions.Internals;
+using Xamarin.Forms;
+
+namespace Xamarin.CommunityToolkit.Converters
+{
+	public class MultiMathExpressionConverter : MultiValueConverterExtension, IMultiValueConverter
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/MultiMathExpressionConverter.shared.cs:10]: what is the purpose of this converter? Is it the same as a MathExpressionConverter but supposed to have several parameters? 

if yes, I believe, we can leave only MathExpressionConverter (where we can check if value an array or a single number)

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathExpression.shared.cs [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace Xamarin.CommunityToolkit.Core
+{
+	internal class MathExpression
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathExpression.shared.cs:11]:
```suggestion
	internal sealed class MathExpression
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathOperator.shared.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+{
+	internal class MathOperator
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Core/Math/MathOperator.shared.cs:5]:
```suggestion
	internal sealed class MathOperator
```

File: samples/XCT.Sample/Xamarin.CommunityToolkit.Sample.csproj [hunk 10]
```diff
@@ -22,6 +22,9 @@
   </ItemGroup>
   
   <ItemGroup>
+    <Compile Update="Pages\Converters\MathExpressionConverterPage.xaml.cs">
+      <DependentUpon>MathExpressionConverterPage.xaml</DependentUpon>
+    </Compile>
```
Comment (AndreiMisiukevich) [inline samples/XCT.Sample/Xamarin.CommunityToolkit.Sample.csproj:27]: Should be removed, I believe)

**Review by AndreiMisiukevich — COMMENTED** [review]
Great job) Very well.
Just a few comments.
Would be great if you could address them all)

**Comment by sattew** [comment]
> Would be great if you could address them all)

@AndreiMisiukevich just one question) 
What does it mean?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/MultiMathExpressionConverter.shared.cs (hunk 7 above)
Comment (sattew) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/MultiMathExpressionConverter.shared.cs:10]: yes, it supposed to have several params and to use MultiBinding. And I think MultiBinding is more comfortable to use in XAML. For example if developer create the function who depends on Height and Width of element and you can't set the bindings without MultiBinding, I guess.

File: samples/XCT.Sample/Xamarin.CommunityToolkit.Sample.csproj (hunk 10 above)
Comment (sattew) [reply samples/XCT.Sample/Xamarin.CommunityToolkit.Sample.csproj]: Sure)

**Comment by sattew** [comment]
```
 C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets#L131

C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.targets(131,5): Error : Failed to retrieve information about 'Microsoft.Net.UWPCoreRuntimeSdk' from remote source 'https://api.nuget.org/v3-flatcontainer/microsoft.net.uwpcoreruntimesdk/index.json'.
  The HTTP request to 'GET https://api.nuget.org/v3-flatcontainer/microsoft.net.uwpcoreruntimesdk/index.json' has timed out after 100000ms.
```

Did I make this error or not?

**Comment by AndreiMisiukevich** [comment]
> Would be great if you could address them all)

https://english.stackexchange.com/questions/181460/what-does-address-an-issue-mean

:) Hope it's clear now

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/MathExpressionConverter/MathExpressionConverter.shared.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+		/// <param name="targetType">The type of the binding target property. This is not implemented.</param>
+		/// <param name="parameter">The expression to calculate.</param>
+		/// <param name="culture">The culture to use in the converter. This is not implemented.</param>
+		/// <returns>A <see cref="double"/> The result of calculating an expression.</returns>
+		public object? Convert(object? value, Type? targetType, object? parameter, CultureInfo culture)
+		{
+			if (parameter is not string expression)
+				throw new ArgumentException("The parameter should be of type String");
+
+			if (value is null || !double.TryParse(value.ToString(), out var xValue))
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/MathExpressionConverter/MathExpressionConverter.shared.cs:23]: We don't use `pattern matching` for checking null.
```suggestion
			if (value == null || !double.TryParse(value.ToString(), out var xValue))
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/MathExpressionConverter/MultiMathExpressionConverter.shared.cs [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+		/// <param name="targetType">The type of the binding target property. This is not implemented.</param>
+		/// <param name="parameter">The expression to calculate.</param>
+		/// <param name="culture">The culture to use in the converter. This is not implemented.</param>
+		/// <returns>A <see cref="double"/> The result of calculating an expression.</returns>
+		public object? Convert(object[]? values, Type? targetType, object? parameter, CultureInfo culture)
+		{
+			if (parameter is not string expression)
+				throw new ArgumentException("The parameter should be of type String");
+
+			if (values is null)
```
Comment (pictos) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Converters/MathExpressionConverter/MultiMathExpressionConverter.shared.cs:24]: Can you change others?
```suggestion
			if (values == null)
```

**Review by pictos — CHANGES_REQUESTED** [review]
Hey there, thanks for this PR. I didn't look too deep into the code, but I just saw these minimal details.

**Comment by pictos** [comment]
@sattew can you open a PR into the docs repo about this feature?

**Comment by sattew** [comment]
> @sattew can you open a PR into the docs repo about this feature?

My english is not well but here PR [#111](https://github.com/MicrosoftDocs/xamarin-communitytoolkit/pull/111)

**Comment by pictos** [comment]
@sattew can you solve this conflict?

**Comment by pictos** [comment]
@sattew FYI I'll do a final review today or tomorrow and if everything looks good I'll approve this puppy! Thanks for you hard work here <3

**Review by pictos — APPROVED** [review]

**Pull request merged**